tools/toolchain/optimized_clang.sh

```diff
@@ -0,0 +1,113 @@
+#!/bin/bash -uex
+
+if [[ -z "${CLANG_BUILD}" ]]; then
+    echo "CLANG_BUILD not specified"
+    exit 1
+elif [[ "${CLANG_BUILD}" = "SKIP" ]]; then
+    echo "Skip building optimized clang"
+    exit 0
+elif [[ "${CLANG_BUILD}" = "INSTALL" ]]; then
+    echo "build and install optimized clang"
+elif [[ "${CLANG_BUILD}" = "INSTALL_FROM" ]]; then
+    echo "install prebuilt optimized clang"
+else
+    echo "Not sure what to do with ${CLANG_BUILD}"
+    exit 1
+fi
+
+if [[ "${CLANG_BUILD}" = "INSTALL" ]] || [[ "${CLANG_BUILD}" = "INSTALL_FROM" ]]; then
+    if [[ -z "${CLANG_ARCHIVE}" ]]; then
+        echo "CLANG_ARCHIVE does not specified"
+        exit 1
+    fi
+fi
+
+ARCH="$(arch)"
+LLVM_EXPERIMENTAL_TARGETS_TO_BUILD="WebAssembly"
+if [[ "${ARCH}" = "x86_64" ]]; then
+    LLVM_TARGET_ARCH=X86
+elif [[ "${ARCH}" = "aarch64" ]]; then
+    LLVM_TARGET_ARCH=AArch64
+    LLVM_EXPERIMENTAL_TARGETS_TO_BUILD="${LLVM_EXPERIMENTAL_TARGETS_TO_BUILD};AArch64"
+else
+    echo "Unsupported architecture: ${ARCH}"
+    exit 1
+fi
+
+SCYLLA_DIR=/mnt
+CLANG_ROOT_DIR="${SCYLLA_DIR}"/clang_build
+CLANG_CHECKOUT_NAME=llvm-project
+CLANG_BUILD_DIR="${CLANG_ROOT_DIR}"/"${CLANG_CHECKOUT_NAME}"
+CLANG_BIN="${CLANG_BUILD_DIR}"/build/bin
+
+# Which LLVM release to build in order to compile Scylla
+LLVM_CLANG_TAG=16.0.6
+CLANG_SUFFIX=16
+
+CLANG_ARCHIVE=$(cd "${SCYLLA_DIR}" && realpath -m "${CLANG_ARCHIVE}")
+
+CLANG_OPTS=(-DCMAKE_C_COMPILER="/usr/bin/clang" -DCMAKE_CXX_COMPILER="/usr/bin/clang++" -DLLVM_USE_LINKER="/usr/bin/ld.lld" -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="${LLVM_TARGET_ARCH};WebAssembly" -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD="${LLVM_EXPERIMENTAL_TARGETS_TO_BUILD}" -DLLVM_TARGET_ARCH="${LLVM_TARGET_ARCH}" -G Ninja -DLLVM_INCLUDE_BENCHMARKS=OFF -DLLVM_INCLUDE_EXAMPLES=OFF -DLLVM_INCLUDE_TESTS=OFF -DLLVM_ENABLE_BINDINGS=OFF -DLLVM_ENABLE_PROJECTS="clang;lld" -DLLVM_ENABLE_RUNTIMES="compiler-rt" -DLLVM_ENABLE_LTO=Thin -DCLANG_DEFAULT_PIE_ON_LINUX=OFF -DLLVM_BUILD_TOOLS=OFF -DCMAKE_C_FLAGS="-Xclang -mllvm -Xclang -vp-counters-per-site=16" -DCMAKE_CXX_FLAGS="-Xclang -mllvm -Xclang -vp-counters-per-site=16")
+SCYLLA_OPTS=(--date-stamp "$(date "+%Y%m%d")" --debuginfo 1 --tests-debuginfo 1 --c-compiler="${CLANG_BIN}/clang" --compiler="${CLANG_BIN}/clang++" --build-dir="build_profile" --out="build_profile.ninja")
```

Comment:
i am surprised that we need to repeat the same compiling options in `CMAKE_CXX_FLAGS` . IIUC, `CXXFLAGS` is for compiling C++ source files, while CFLAGS is for compiling options in general. despite that CMake [document](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS.html) puts that

> Language-wide flags for language <LANG> used when building for all configurations

in practice, `CMAKE_C_FLAGS` is used for all "C"ompiling options. not just for compiling the C sources. in other words, all flags in `CMAKE_C_FLAGS` are applied when compiling C++ sources as well. that's why we always set, for instance, `-Werror -O2`, in `CMAKE_C_FLAGS` instead of in both of these two CMake variables.

Reply:
also, note to myself, instead of specifying the path to the generated `build.ninja` file, we could just use `cmake` for generating the `build.ninja` when we migrate to CMake building system.

Reply:
I added this because we get tons of following warning during clang build:
`LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.`
However, I just realized that the warning still appear on build log both on new version and old version (both DCMAKE_C_FLAGS and DCMAKE_CXX_FLAGS or DCMAKE_C_FLAGS only).

So the option actually does not fix the problem, probably we need to find other solution, or just ignore it.

Reply:
sorry. i just checked the generated `build.ninja`, and proved that i was wrong -- if we add the said flags to `CMAKE_C_FLAGS`, only C source files would use those flags. but if we add them to `CMAKE_CXX_FLAGS`, these flags apply to C++ source files.

Reply:
> However, I just realized that the warning still appear on build log both on new version and old version (both DCMAKE_C_FLAGS and DCMAKE_CXX_FLAGS or DCMAKE_C_FLAGS only).
> 
> So the option actually does not fix the problem, probably we need to find other solution, or just ignore it.

i am running `tools/toolchain/optimized_clang.sh` in hope to reproduce this issue.

Reply:
Tested this ( -DLLVM_VP_COUNTERS_PER_SITE=16), but still getting same warning while running `clang-tblgen`.

```
[603/3454] cd /mnt/clang_build/llvm-project/build && /mnt/clang_build/llvm-project/build/bin/clang-tblgen -gen-clang-diags-index-name -I /mnt/clang_build/llvm-project/clang/include/clang/Basic -I/mnt/clang_build/llvm-project/clang/include -I/mnt/clang_build/llvm-project/build/tools/clang/include -I/mnt/clang_build/llvm-project/build/include -I/mnt/clang_build/llvm-project/llvm/include /mnt/clang_build/llvm-project/clang/include/clang/Basic/Diagnostic.td --write-if-changed -o tools/clang/include/clang/Basic/DiagnosticIndexName.inc -d tools/clang/include/clang/Basic/DiagnosticIndexName.inc.d
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
LLVM Profile Warning: Unable to track new values: Running out of static counters.  Consider using option -mllvm -vp-counters-per-site=<n> to allocate more value profile counters at compile time.
```

BTW, specified parameter (16) is just taken from other project which trying to build PGO clang, but I don't really understand which size we should assign here.

Reply:
> Tested this ( -DLLVM_VP_COUNTERS_PER_SITE=16), but still getting same warning while running `clang-tblgen`.

I checked verbose ninja build log (-v), LLVM_VP_COUNTERS_PER_SITE does not append `-Xclang -mllvm -Xclang -vp-counters-per-site=<n>` to the compilation arguments for some reason.
Make be related with the if else condition of HandleLLVMOptions.cmake, but not sure what is missing.

Reply:
I don't find a way to solve this. Probably we can fix this later, after the patch merged.